### PR TITLE
feat: update webpack-dev-server types to match 4.5

### PR DESCRIPTION
**Pull request by alan-agius4**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/webpack/webpack-dev-server/blob/master/CHANGELOG.md#450-2021-11-13 and https://github.com/webpack/webpack-dev-server/blob/master/lib/options.json
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/webpack-dev-server/index.d.ts [hunk 1]
```diff
@@ -270,8 +274,20 @@ declare namespace WebpackDevServer {
...
+            | { type?: string | boolean | WebSockerType; options?: Record<string, any> };
+
+        /**
+         * Allows to set server and options (by default 'http')."
+         */
+        server?: string | ServerType | { type?: ServerType; options?: https.ServerOptions };
     }
+
+    type ServerType = 'http' | 'https' | 'spdy';
+    type WebSockerType = 'sockjs' | 'ws';
```
Comment (AviVahl) [inline types/webpack-dev-server/index.d.ts:290]: Typo? (Sucker->Socket)

Comment (alan-agius4) [reply types/webpack-dev-server/index.d.ts:290]: Yeah :P

File: types/webpack-dev-server/index.d.ts [hunk 2]
```diff
@@ -270,8 +274,20 @@ declare namespace WebpackDevServer {
...
+            | { type?: string | boolean | WebSocketType; options?: Record<string, any> };
+
+        /**
+         * Allows to set server and options (by default 'http')."
+         */
+        server?: string | ServerType | { type?: ServerType; options?: https.ServerOptions };
     }
+
+    type ServerType = 'http' | 'https' | 'spdy';
+    type WebSocketType = 'sockjs' | 'ws';
```
Comment (lex111) [inline types/webpack-dev-server/index.d.ts:290]: For consistency, please define these new types above, before `Configuration` interface.

File: types/webpack-dev-server/webpack-dev-server-tests.ts [hunk 3]
```diff
@@ -273,6 +273,39 @@ server = new WebpackDevServer(
...
+                passphrase: '----',
+            },
+        },
+    },
+    compiler,
+);
+
+server = new WebpackDevServer(
+    {
+        server: 'spdy',
```
Comment (lex111) [inline types/webpack-dev-server/webpack-dev-server-tests.ts:288]: Please put the new options in the existing webpack dev server configs above (`c2`--`c6`).

**Comment by alan-agius4** [comment]
@lex111 PTAL.

File: types/webpack-dev-server/index.d.ts [hunk 4]
```diff
@@ -193,11 +196,15 @@ declare namespace WebpackDevServer {
          * Serve over HTTP/2 using spdy. This option is ignored for Node 10.0.0
          * and above, as spdy is broken for those versions. The dev server will
          * migrate over to Node's built-in HTTP/2 once Express supports it.
+         *
+         * @deprecated Use the 'server' option."
```
Comment (lex111) [inline types/webpack-dev-server/index.d.ts:200]:
```suggestion
         * @deprecated Use the 'server' option.
```

File: types/webpack-dev-server/index.d.ts [hunk 5]
```diff
@@ -193,11 +196,15 @@ declare namespace WebpackDevServer {
...
          */
         http2?: boolean | undefined;
         /**
          * By default dev-server will be served over HTTP. It can optionally be
          * served over HTTP/2 with HTTPS.
+         *
+         * @deprecated Use the 'server' option."
```
Comment (lex111) [inline types/webpack-dev-server/index.d.ts:207]:
```suggestion
         * @deprecated Use the 'server' option.
```

File: types/webpack-dev-server/index.d.ts [hunk 6]
```diff
@@ -270,7 +277,16 @@ declare namespace WebpackDevServer {
...
          */
-        webSocketServer?: boolean | 'sockjs' | 'ws' | string | (() => void) | Record<string, any>;
+        webSocketServer?:
+            | string
+            | boolean
+            | WebSocketType
+            | { type?: string | boolean | WebSocketType; options?: Record<string, any> };
+
+        /**
+         * Allows to set server and options (by default 'http')."
```
Comment (lex111) [inline types/webpack-dev-server/index.d.ts:287]:
```suggestion
         * Allows to set server and options (by default 'http').
```

**Review by lex111 — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by alan-agius4** [comment]
Ready to merge

**Pull request merged**